File: backend/app/routers/subcontract_valuations.py
"""Subcontract valuations router — Chat 35 §R4.1 (Prompt 2.8b).

Mounted under /api/v1 via `server.py`.

Endpoints:
  POST   /subcontract-valuations                          .create  → 201
  GET    /subcontract-valuations?subcontract_id=&status=  .view
  GET    /subcontract-valuations/{id}                     .view
  POST   /subcontract-valuations/{id}/submit              .create
  POST   /subcontract-valuations/{id}/certify             .certify
  POST   /subcontract-valuations/{id}/reject              .certify

Error mapping:
  - ValuationNotFoundError → 404 (cross-tenant or missing)
  - ValuationStateError    → 409 (bad workflow transition)
  - ValueError             → 422 (payload validation)
"""
from __future__ import annotations

import uuid
from datetime import date
from typing import Any, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, Request
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.auth.deps import get_current_user, require_permission
from app.auth.permissions import UserPermissions
from app.db import get_db
from app.models.user import User
from app.services import subcontract_valuations as svc
from app.services.subcontract_valuations import (
    ValuationNotFoundError, ValuationStateError,
)


router = APIRouter(tags=["subcontract_valuations"])


# ---------------------------------------------------------------------------
# Schemas
# ---------------------------------------------------------------------------

class ValuationCreateBody(BaseModel):
    subcontract_id: uuid.UUID
    gross_applied_to_date: str
    labour_portion: str = "0"
    materials_portion: str = "0"
    period_start: Optional[date] = None
    period_end: Optional[date] = None


class ValuationCertifyBody(BaseModel):
    transaction_date: Optional[date] = None
    description: Optional[str] = Field(None, max_length=500)
    # Chat 39 §R2 A4: required. No silent LIMIT-1 guess on the project.
    # Caller MUST specify which budget line bears the subcontractor cost.
    budget_line_id: uuid.UUID


class ValuationRejectBody(BaseModel):
    reason: str = Field(..., min_length=1, max_length=2000)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _map(exc: Exception) -> HTTPException:
    if isinstance(exc, ValuationNotFoundError):
        return HTTPException(status_code=404, detail=str(exc) or "Not found")
    if isinstance(exc, ValuationStateError):
        return HTTPException(status_code=409, detail=str(exc))
    if isinstance(exc, ValueError):
        return HTTPException(status_code=422, detail=str(exc))
    return HTTPException(status_code=500, detail=str(exc))


def _serialise(v, *, perms: UserPermissions) -> dict[str, Any]:
    has_sensitive = perms.has("subcontract_valuations.view_sensitive")
    return svc.serialise(v, include_sensitive=has_sensitive)


# ---------------------------------------------------------------------------
# Routes
# ---------------------------------------------------------------------------

@router.post("/subcontract-valuations", status_code=201)
def create_valuation_endpoint(
    body: ValuationCreateBody,
    request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(
        require_permission("subcontract_valuations.create")
    ),
    db: Session = Depends(get_db),
):
    try:
        v = svc.create_valuation(
            db,
            subcontract_id=body.subcontract_id,
            gross_applied_to_date=body.gross_applied_to_date,
            labour_portion=body.labour_portion,
            materials_portion=body.materials_portion,
            period_start=body.period_start,
            period_end=body.period_end,
            user=current, perms=perms, request=request,
        )
    except (ValuationNotFoundError, ValuationStateError, ValueError) as exc:
        raise _map(exc)
    db.commit()
    db.refresh(v)
    return _serialise(v, perms=perms)


@router.get("/subcontract-valuations")
def list_valuations_endpoint(
    subcontract_id: Optional[uuid.UUID] = Query(None),
    status: Optional[str] = Query(None),
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(
        require_permission("subcontract_valuations.view")
    ),
    db: Session = Depends(get_db),
):
    try:
        rows = svc.list_valuations(
            db, user=current, perms=perms,
            subcontract_id=subcontract_id, status=status,
            limit=limit, offset=offset,
        )
    except (ValuationNotFoundError, ValueError) as exc:
        raise _map(exc)
    items = [_serialise(r, perms=perms) for r in rows]
    return {"items": items, "total": len(items)}


@router.get("/subcontract-valuations/{val_id}")
def get_valuation_endpoint(
    val_id: uuid.UUID,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(
        require_permission("subcontract_valuations.view")
    ),
    db: Session = Depends(get_db),
):
    try:
        v = svc.get_valuation(db, val_id, user=current, perms=perms)
    except ValuationNotFoundError as exc:
        raise _map(exc)
    return _serialise(v, perms=perms)


@router.post("/subcontract-valuations/{val_id}/submit")
def submit_valuation_endpoint(
    val_id: uuid.UUID,
    request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(
        require_permission("subcontract_valuations.create")
    ),
    db: Session = Depends(get_db),
):
    try:
        v = svc.submit_valuation(
            db, val_id, user=current, perms=perms, request=request,
        )
    except (ValuationNotFoundError, ValuationStateError, ValueError) as exc:
        raise _map(exc)
    db.commit()
    db.refresh(v)
    return _serialise(v, perms=perms)


@router.post("/subcontract-valuations/{val_id}/certify")
def certify_valuation_endpoint(
    val_id: uuid.UUID,
    body: ValuationCertifyBody,
    request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(
        require_permission("subcontract_valuations.certify")
    ),
    db: Session = Depends(get_db),
):
    try:
        v = svc.certify_valuation(
            db, val_id, user=current, perms=perms,
            transaction_date=body.transaction_date,
            description=body.description,
            budget_line_id=body.budget_line_id,
            request=request,
        )
    except (ValuationNotFoundError, ValuationStateError, ValueError) as exc:
        raise _map(exc)
    db.commit()
    db.refresh(v)
    return _serialise(v, perms=perms)


@router.post("/subcontract-valuations/{val_id}/reject")
def reject_valuation_endpoint(
    val_id: uuid.UUID,
    body: ValuationRejectBody,
    request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(
        require_permission("subcontract_valuations.certify")
    ),
    db: Session = Depends(get_db),
):
    try:
        v = svc.reject_valuation(
            db, val_id, user=current, perms=perms,
            reason=body.reason, request=request,
        )
    except (ValuationNotFoundError, ValuationStateError, ValueError) as exc:
        raise _map(exc)
    db.commit()
    db.refresh(v)
    return _serialise(v, perms=perms)

File: frontend/src/lib/api/subcontractValuations.js
/**
 * Subcontract Valuations API client — Chat 48 (Build Pack 2.8-FE-ii §R3.1).
 *
 * Thin axios wrappers around the `/v1/subcontract-valuations` endpoints.
 * Mirrors `lib/api/subcontracts.js` conventions: shared `lib/api.js`
 * baseURL is `/api`, so paths here are prefixed `/v1/...` ourselves.
 * All functions return the response body directly (`data`).
 *
 * Endpoint contract (verified origin/main, Chat 48,
 * routers/subcontract_valuations.py + services/subcontract_valuations.py
 * + models/sc_valuations.py — alembic head 0043_document_folders):
 *
 *   POST   /v1/subcontract-valuations              → 201 valuation
 *   GET    /v1/subcontract-valuations              → { items, total }
 *      params: subcontract_id, status, limit, offset
 *   GET    /v1/subcontract-valuations/{id}         → valuation
 *   POST   /v1/subcontract-valuations/{id}/submit  → updated valuation
 *   POST   /v1/subcontract-valuations/{id}/certify → updated valuation
 *   POST   /v1/subcontract-valuations/{id}/reject  → updated valuation
 *
 * Money fields are STRINGS on the wire (Pydantic Decimal serialisation).
 * The UI must NEVER do float maths — the server is the single source of
 * truth for all money calculations (retention, CIS, net payable). This
 * client is a thin pass-through: it does not validate, transform, or
 * compute. Callers send strings, callers receive strings.
 *
 * Error mapping (mirrored in hook callers + components):
 *   - 404  not found / cross-tenant (never leaks existence).
 *   - 409  STATE error (wrong workflow status, wrong parent status,
 *          missing budget_line_id at certify, budget line not on the
 *          subcontract's project). UI surfaces `detail` distinctly from
 *          422 and refetches to resync.
 *   - 422  VALIDATION error (gross < previously certified, labour +
 *          materials != gross_this_cert, net payable < 0, negative
 *          inputs at create, blank rejection reason). UI surfaces the
 *          server `detail` verbatim — the maths messages are
 *          user-meaningful.
 *   - 403  missing permission.
 *
 * Permission map (verified, differs from earlier guesses):
 *   - submit  → `subcontract_valuations.create`  (NOT a separate perm)
 *   - certify → `subcontract_valuations.certify`
 *   - reject  → `subcontract_valuations.certify` (shared)
 *
 * Sensitive money fields on the GET serialiser (gated by
 * `subcontract_valuations.view_sensitive`, null otherwise):
 *   previous_certified_net, retention_this_cert, cis_rate_pct,
 *   cis_deduction_this_cert, net_payable_this_cert
 */
import { api } from '@/lib/api';


// ─── List + read ────────────────────────────────────────────────────

/**
 * List valuations visible to the caller.
 *
 * Params (all optional except convention):
 *   - subcontractId — REQUIRED in practice (this surface is always
 *     scoped to a parent subcontract); sent as `subcontract_id`.
 *   - status        — 'Draft' | 'Submitted' | 'Certified' | 'Rejected'.
 *                     Other values 422 server-side.
 *   - limit         — 1..200 (server-clamped).
 *   - offset        — >= 0.
 *
 * Snake_case forwarding is hard-pinned in __tests__/subcontractValuations.test.js
 * — the backend rejects camelCase keys.
 */
export async function listValuations(
  { subcontractId, status, limit, offset, signal } = {},
) {
  const params = {};
  if (subcontractId != null) params.subcontract_id = subcontractId;
  if (status != null) params.status = status;
  if (limit != null) params.limit = limit;
  if (offset != null) params.offset = offset;
  const { data } = await api.get(
    '/v1/subcontract-valuations', { signal, params },
  );
  return data;
}

export async function getValuation(valuationId, { signal } = {}) {
  const { data } = await api.get(
    `/v1/subcontract-valuations/${valuationId}`, { signal },
  );
  return data;
}


// ─── Create ──────────────────────────────────────────────────────────

/**
 * Create a Draft valuation. Body (verified ValuationCreateBody):
 *   { subcontract_id          (uuid, required),
 *     gross_applied_to_date   (str Decimal, required),
 *     labour_portion          (str Decimal, default "0"),
 *     materials_portion       (str Decimal, default "0"),
 *     period_start            (date | null),
 *     period_end              (date | null) }
 *
 * 422 from the backend at create:
 *   - gross_applied_to_date < 0
 *   - labour_portion < 0 or materials_portion < 0
 *
 * 409 from the backend at create:
 *   - parent subcontract not in {Active, Completed}
 *     ("Cannot create a valuation on a {status} subcontract …")
 *
 * Thin pass-through: caller is responsible for sending money as
 * STRINGS (e.g. "10000.00", not 10000). The wire test asserts this.
 */
export async function createValuation(body) {
  const { data } = await api.post('/v1/subcontract-valuations', body);
  return data;
}


// ─── Lifecycle transitions ──────────────────────────────────────────
//
// All action endpoints return the updated valuation. On 409 (wrong
// status), the caller surfaces the server detail and refetches to
// resync the displayed status badge.

/**
 * Submit: Draft → Submitted.
 * Body: {} (empty). Permission: subcontract_valuations.create.
 * 409 if status != Draft.
 */
export async function submitValuation(valuationId) {
  const { data } = await api.post(
    `/v1/subcontract-valuations/${valuationId}/submit`, {},
  );
  return data;
}

/**
 * Certify: Submitted → Certified. Body (verified ValuationCertifyBody):
 *   { transaction_date  (date | null),
 *     description       (str | null, <= 500),
 *     budget_line_id    (uuid, REQUIRED) }
 *
 * Permission: subcontract_valuations.certify.
 *
 * The backend REFUSES to guess the budget line — `budget_line_id` is
 * a hard requirement. This client is a pass-through; the UI's certify
 * dialog disables Confirm until a line is selected, and the wire test
 * asserts the body shape.
 *
 * 409 cases:
 *   - status != Submitted ("must be Submitted")
 *   - missing budget_line_id  ("budget_line_id is required …")
 *   - budget line not on this subcontract's project
 *     ("budget_line is not on the subcontract's project")
 *
 * 422 cases (the maths guards — surface verbatim):
 *   - gross_this_cert < 0  (application went backwards)
 *   - labour + materials != gross_this_cert  (split mismatches the
 *     NEW-this-cert gross, NOT the cumulative — see §R4.4 UX note)
 *   - net_payable_this_cert < 0
 */
export async function certifyValuation(valuationId, body) {
  const { data } = await api.post(
    `/v1/subcontract-valuations/${valuationId}/certify`, body,
  );
  return data;
}

/**
 * Reject: Submitted → Rejected. Body: { reason (str, 1..2000) }.
 * Permission: subcontract_valuations.certify (shared with certify).
 *
 * 409 if status != Submitted.
 * 422 if reason blank ("rejection reason is required").
 */
export async function rejectValuation(valuationId, body) {
  const { data } = await api.post(
    `/v1/subcontract-valuations/${valuationId}/reject`, body,
  );
  return data;
}
